docs/dev/guidelines/criteria-builder.rst

```diff
@@ -0,0 +1,294 @@
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+        	Predicate one = criteriaBuilder.equal(x, z);
+        	Predicate two = criteriaBuilder.notEqual(a, b);
+
+        	return criteriaBuilder.and(one, two, ...);
+        };
+
+- **OR:** We can perform the `or` operation on an arbitrary number of predicates via the criteriaBuilder object, which results in a new Predicate.
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+            Predicate one = criteriaBuilder.equal(x, z);
+            Predicate two = criteriaBuilder.notEqual(a, b);
+
+            return criteriaBuilder.or(one, two, ...);
+        };
+
+- **EQUAL / NOT EQUAL:**
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+         	Predicate one = criteriaBuilder.equal(root.get(User_.IS_INTERNAL), true);
+        	Predicate two = criteriaBuilder.notEqual(root.get(User_.ACTIVATED), true);
+
+        	return criteriaBuilder.and(one, two, ...);
+        };
+
+- **NOT:**
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+            ...
+            Predicate predicate = criteriaBuilder.exists(subQuery).not();
+
+            return criteriaBuilder.equals(predicate);
+        };
+
+- **IN:** To check if the collection contains a value.
+
+    .. code-block:: java
+
+        return (root, query, criteriaBuilder) -> {
+         	Predicate in = criteriaBuilder.in(root.get(User_.ID)).value(ids);
+            return in;
+        };
+
+
+4. Joins
+========
+
+Many different joins are available (e.g. Join, ListJoin, SetJoin, CollectionJoin, ...) - please choose the right one.
+
+- If we want to join from X to Y, we need to define the column and the join type. Please mind that when the join type is not specified, Inner Join is made by default.
+
+    .. code-block:: java
+
+        Join<X, Y> join = root.join(X_.COLUMN, JoinType.LEFT);
+
+- We can define custom on clauses to specify the join condition.
+
+    .. code-block:: java
+
+        Join<X, Y> join = root.join(X_.COLUMN, JoinType.LEFT);
+        join.on(criteriaBuilder.in(join.get(Y_.NAME)).value(names));
+
+- We can concatenate joins.
+
+    .. code-block:: java
+
+        Join<X, Z> join = root.join(X_.COLUMN, JoinType.LEFT).join(Y_.COLUMN, JoinType.LEFT);
+
+
+4. Sub-Queries
+==============
+
+Sub-queries are usually fine unless they are dependent sub-queries (also known as `correlated <https://en.wikipedia.org/wiki/Correlated_subquery>`_ sub queries).
+
+1. **Dependent Sub-Query:**
+    In a SQL database query, a correlated sub-query is a sub-query (a query nested inside another query) that uses values from the outer query. But with a dependent sub-query you might run into performance problems because a dependent sub-query typically needs to be run once for each row in the outer query, e.g. if your outer query has 1000 rows, the sub-query will be run 1000 times.
```

Comment:
```suggestion
    In an SQL database query, a correlated sub-query is a sub-query (a query nested inside another query) that uses values from the outer query. But with a dependent sub-query you might run into performance problems because a dependent sub-query typically needs to be run once for each row in the outer query, e.g. if your outer query has 1000 rows, the sub-query will be run 1000 times.
```
Nitpick on S-Q-L vs sequel